numpy/lib/function_base.py

```diff
@@ -3929,21 +3928,33 @@ def _quantile_ureduce_func(a, q, axis=None, out=None, overwrite_input=False,
             indices_above = indices_above[:-1]
             n = np.isnan(ap[-1:, ...])
 
-        x1 = take(ap, indices_below, axis=axis) * weights_below
-        x2 = take(ap, indices_above, axis=axis) * weights_above
+        x1 = take(ap, indices_below, axis=axis)
+        x2 = take(ap, indices_above, axis=axis)
+
+        # the linear interpolation below ensure monotonicity with floating
+        # point operations and was proposed in:
+        # https://math.stackexchange.com/a/1798323 by Pedro Gimeno
+        diff_x2_x1 = x2 - x1
 
-        # ensure axis with q-th is first
         x1 = np.moveaxis(x1, axis, 0)
         x2 = np.moveaxis(x2, axis, 0)
 
+        r_above = np.add(x1, diff_x2_x1 * weights_above, out=out)
+        r_below = np.subtract(
+            x2, diff_x2_x1 * weights_below, out=r_above,
+            where=weights_above < 0.5
+        )
+
         if zerod:
-            x1 = x1.squeeze(0)
-            x2 = x2.squeeze(0)
+            r_above = r_above.squeeze(0)
 
         if out is not None:
-            r = add(x1, x2, out=out)
+            out[...] = r_above
+            r = out
```

Comment:
Assignment to out should not be necessary here, the above operation is already in-place (i.e. similar to the old branch, which did not do this).

Reply:
@seberg is the double-assignment cleanup a blocker?

Reply:
Not on its own, but I think the above also needs some cleanups, maybe I should just put it on my todo list to do it myself.